### ISSUE TITLE
sails-mongo: Date objects turn into string

### ISSUE THREAD

**Issue by sailsbot** [balderdashy/sails#4020]
_From @Parziphal on August 25, 2015 18:46_

Hello. I've found this strange behaviour:

I create a document like this:

``` js
User.create({
  conversations: [
    {
      messages: [
        {
          createdAt: new Date()
        }
      ]
    }
  ]
}).exec(function() {});
```

Using Robomongo, I check the new document and I see that the `conversations.messages.createdAt` field is a Date object, so everything's fine. But then I append a new object to the `messages` array like this:

``` js
User.find({}).limit(1).exec(function(err, users) {
  users[0].conversations[0].messages.push({createdAt: new Date()});
  users[0].save();
});
```

And suddenly, all `conversations.messages.createdAt` fields turn into `string` objects. Is this an intended behaviour? Am I missing something? Maybe this will be fixed when this feature is added https://github.com/balderdashy/sails-mongo/issues/44 ?

**EDIT**

Actually any `save` operation screws the data, turning Date objects into strings.

_Copied from original issue: balderdashy/sails-mongo#325_

**Comment by sailsbot** [comment]
Thanks for posting, @Parziphal.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
_From @devinivy on September 25, 2015 10:43_

What version of waterline and the adapter are you using?

**Comment by sailsbot** [comment]
_From @Parziphal on September 26, 2015 5:21_

`npm list -g` says:
waterline@0.10.26

**Comment by sailsbot** [comment]
Thanks for posting, @Parziphal.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:
- review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
- create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sailsbot** [comment]
_From @obj63mc on December 9, 2015 14:14_

Just wondering if there is any update on this, major bug for example - 

```
Post.findOne({postid:req.query.postid}).exec(function(err, p){
    p.postdate = new Date();
    p.save();
});
```

This is always saving the date as a string not an ISOdate in mongo.  Any ideas how to work around it?  I am using sails-mongo 0.11.5 and sails 0.11.3

**Comment by sailsbot** [comment]
_From @devinivy on December 14, 2015 14:54_

I don't use this adapter, and I haven't had time to look into it.  Do the default `createdAt` and `updatedAt` fields have this problem?

**Comment by sailsbot** [comment]
_From @obj63mc on December 14, 2015 15:14_

The created and updated work as expected, it is only when you are using a schemaless type object and setting a date field somewhere in the object.  This did not used to happen in older versions though.  Not sure which versions but I know for sure it worked in 0.9x

**Comment by sailsbot** [comment]
_From @particlebanana on January 25, 2016 19:39_

I tried the examples in this issue with the latest version of the adapter and didn't have any issues. Can someone post a link to a repo that reproduces this issue so I can get to the bottom of it?

Specifically can I see model definitions and a set of queries I can run that shows the error?

**Comment by sailsbot** [comment]
_From @time2try on June 16, 2016 18:50_

@particlebanana In reference to your comment I have added the following to demonstrate how our team is facing the same issues. I hope this helps you reproduce this issue.

**MODEL:** Twitter.js

``` javascript
module.exports = {
  tableName : 'Twitter'
};
```

**CONTROLLER:** TweetController.js

``` javascript
replySent: function(req, res){
    var params = req.params.all();
    var tweetID = params.tweetID;
    var reply = params.reply;

    var replyBlock = {
      date_replied: new Date(),
      content: reply
    };
        Twitter.findOne({'tweet_id': tweetID})
          .exec(function (err, tweet) {
            if (err) {
              return res.serverError(err);
            } else {
              if(_.isEmpty(tweet.reply)){
                tweet.reply= [replyBlock];
              }else{
                tweet.reply.push(replyBlock);
              }
              tweet.save(function(err){
                if(err) return res.serverError(err);
                return res.ok('Success!');
              });
            }
          });
}
```

**Tweet Object** in MongoDB before the `tweet.save(function(err){...})`. Notice the `ISODate()` on `date_published`.

``` bson
{ 
    "_id" : ObjectId("5762d019d432601501bfe083"), 
    "tweet_id" : "743475907659079680", 
    "retweet_count" : 0, 
    "content" : "Wow! Kony, Inc. is ahead of the pack! Most Visionary Leader-Gartner MQ for Mobile Application Development Platforms! <a href=https://t.co/1QvGHxV9k5>https://t.co/1QvGHxV9k5</a> ", 
    "date_published" : ISODate("2016-06-16T16:10:44.000+0000"), 
    "username" : "jeankondo1", 
    "extracted_urls" : [
        "https://t.co/1QvGHxV9k5"
    ], 
    "search_term" : "\"application development\" -job", 
    "analyzed_location" : {
        "province" : "US Southwest", 
        "source" : "Author"
    }
}
```

**Tweet Object** after `tweet.save(function(err){...})`. Notice `date_published` now. The `ISODate()` has been removed and replaced by a `String`.

``` bson
{ 
    "_id" : ObjectId("5762d04cd06c833168bd9851"), 
    "tweet_id" : "743475907659079680", 
    "retweet_count" : NumberInt(0), 
    "content" : "Wow! Kony, Inc. is ahead of the pack! Most Visionary Leader-Gartner MQ for Mobile Application Development Platforms! <a href=https://t.co/1QvGHxV9k5>https://t.co/1QvGHxV9k5</a> ", 
    "date_published" : "2016-06-16T16:10:44.000Z", 
    "username" : "jeankondo1", 
    "extracted_urls" : [
        "https://t.co/1QvGHxV9k5"
    ], 
    "search_term" : "\"application development\" -job",  
    "analyzed_location" : {
        "province" : "US Southwest", 
        "source" : "Author"
    }, 
    "reply" : [
        {
            "date_replied" : "2016-06-16T17:46:54.951Z", 
            "content" : "hello mate"
        }
    ], 
    "updatedAt" : ISODate("2016-06-16T17:46:55.029+0000")
}
```

**Comment by mikermcneil** [comment]
@Parziphal In Sails v1, if you set `type: 'ref'` for an attribute and store a date this way, it should come in and out of Mongo untouched.  You cannot use this for default createdAt or updatedAt timestamps however -- they're standardized and expected either to be a number (JS timestamp; i.e. epoch milliseconds) or a string (JSON timestamp; ISO 8601)

**Comment by sgress454** [comment]
In Sails v0.12.x, the best suggestion would be to not store Javascript date objects in Mongo, but to serialize them first via either `.getTime()` or `.toJSON()`, the results of either of which can be passed to `new Date()` to deserialize them.  In Sails v1, this will be done for you if the attribute is set to the `json` type.

**Comment by luislobo** [comment]
@sgress454 @mikermcneil  actually, one should be able to save whatever you want into a MongoDB attribute, if you don't define the structure, or, for example, if you define it as an array.

In Waterline 0.12.x, if you change in 
https://github.com/balderdashy/waterline/blob/0.12.x/lib/waterline/model/lib/defaultMethods/save.js#L85
from:
```javascript
var data = _.clone(proto.toObject());
```
to
```javascript
var data = _.clone(proto);
```
it just works perfect.
The problem is that toObject is stringifying and converting back to JSON, thus, making date objects string date objects.

I've run Unit tests with that change and all works.
The only thing that fails are integration tests, but those fail even before the change.

Originally, that line wasn't there, it was added to fix an m:m issue and the .add method, (without the _.clone) and later clone was added to remove all functions.

**Comment by luislobo** [comment]
https://github.com/luislobo/test-4020

**Comment by particlebanana** [comment]
@luislobo sounds like an easy solution is to use `update` instead of the `.save()` method 😄 

I'm hesitant to change that 3 year old code when I don't have time to do a deep test and I don't remember why it's there. It's possible that it would be fine but I would need to test it throughly to ensure it's not a breaking change. Waterline doesn't actually support these types of Mongo attributes (only JSON attributes). In Sails 1.0 the `save` method is gone along with nested updates so that is taken care of already in future versions.

**Comment by luislobo** [comment]
@particlebanana sounds good to me! Using Update does workaround the issue.

I was just adding light to the cause and possible fixes of the `save` method. I understand that the energy is now towards Sails 1.0.

Now, about:
> Waterline doesn't actually support these types of Mongo attributes (only JSON attributes). 

According to http://sailsjs.com/documentation/concepts/models-and-orm/attributes, Sails supports `json` and `arrays`.
What do you mean by what you mentioned?

**Comment by particlebanana** [comment]
`json` attributes are values that can be json stringified. In adapters that don't support storing json directly (like MySQL) the values are always stringified first.

**Comment by luislobo** [comment]
Oh! alright, makes sense!

Well, good thing that `save` is gone in Sails 1!

**Comment by galcivar** [comment]
Hi guys,
I am using Sails v1 and get a similar error using mongodb native:
```
db.collection('user').update({
          _id: new ObjectId(req.session.me)
        }, {
          $push: {
            emailChange: {
              requested: {
                IPv6: IPv6,
                IPv4: IPv4,
                location: location,
                timestamp: new Date()
              }
            }
          }
        }, ...
```
timestamp is saved as a string by the adapter, although I am declaring it as a Date Object.

Declaring it in the model with 'ref' is not applicable as you can see it is inside an Array with objects in Mongodb.

What is the correct approach to be able to save it as a date object in Mongodb?

**Comment by sgress454** [comment]
🎉 Closing now that v1.0 is officially released.
